Cucumber test for bootstrapping minion with SSH key

hustodemon opened the pull request:
I also squeezed in a small (but important) fix in this PR.

## What does this PR change?

QA Reviwers: Please review the "Test part" and check if it makes sense.

Code Reviewers: Please review the "Code part".

### Test part


Introduced a (secondary) feature for testing for bootstrapping SLE minion,
highlights:
- added a pre-created ssh key pair
- setup scenario for adding keys to the `authorized_keys` file of the minion,
  cleanup scenario for this
- one scenario tries bootstrapping minion with the key with wrong passphrase
- one scenario tries bootstrapping minion with the key with correct passphrase
- the rest of the scenario tries to bring the minion back to the original state
- there are some questions directly in the file marked with `todo`


### Code part
This decouples the temporary key file **name** creation from creating the file
itself. This allows the clean up code to be executed in more situations
(previously, when the file creation succeeded, but the mode setting failed, the
cleanup code was not executed!).

Introduced one more small change: don't check for the mode of the tmp key file
after changing it via `set_mode`. The value returned was not reliable (it
looks like that some salt versions return "0600", some return "0o600"). Let's
rely on the exceptions thrown from salt in this case.

Finally, an info about the key file lifecycle was added to the UI (see the screenshot).

## GUI diff

After:
![ss](https://user-images.githubusercontent.com/1412268/80300889-a7536900-87a0-11ea-949e-6dfefbee0a98.png)

- [x] **DONE**

## Documentation
- No documentation needed: tests + bugfixes

- [x] **DONE**

## Test coverage
- Cucumber tests were added

- [x] **DONE**

## Links

Tracks https://github.com/SUSE/spacewalk/issues/11114

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)


## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

srbarrios commented on testsuite/features/step_definitions/common_steps.rb:
```diff
@@ -1161,3 +1161,35 @@ def server_secret
+
+# todo is this the right place?
+# common constants for SSH bootstrap with SSH key auth test
+key_filename = 'id_rsa_bootstrap-passphrase_linux.pub'
+## authorized_keys paths on the client
+auth_keys_path = '/root/.ssh/authorized_keys'
+auth_keys_sav_path = '/root/.ssh/authorized_keys.sav'
+
```
I would prefer to have it as part of constants.rb file. But we couldn't arrive at an agreement related to constants in our squad.

Bischoff replied:
I think these constants should be part of the individual step definitions:

1 - if they were real global constants, they should start with `$`
2 - they are too specific to become really global

Please move them into the steps. It's not a big deal if they are repeated a few times.

hustodemon replied:
Moving to steps.

srbarrios commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    Given I am authorized
+    When I go to the minion onboarding page
+    Then I should see a "accepted" text
+    When I navigate to "rhn/systems/Overview.do" page
```
I personally prefer to don't have calls to endpoints in out steps, I don't remember now the step, but we have another that has the same behavior without having the URI there.

Bischoff replied:
100 % agree, we should avoid internals in high-level cucumber text.

hustodemon replied:
Looking at the test codebase, it even seems this seems to be the way to navigate to a "general" page. You are right, that there is another way - it's used in docker building feature (and it's tighly related to it), but the problem is, we'd need to write extra step for each navigation target. Btw, my code was a c&p from an existing bootstrapping test. And while I also don't like having `rhn/blabla/bar` in the steps that much, this seems like this should be done separately as a refactoring that targets all occurences of this step: `When I navigate to "rhn/systems/Overview.do" page`.

```bash
git grep 'I navigate to' | grep  "\.do" | wc -l
> 22

git grep 'I navigate to' | grep -v "\.do" | wc -l
> 11
```

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+
+Feature: Be able to bootstrap a Salt minion via the GUI using SSH key
+
+  Scenario: Delete SLES minion system profile before bootstrap test
```
maybe "before bootstrap with SSH key test" to make it more obvious what bootstrap test it's about.

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    When I backup the SSH authorized_keys file of host "sle_minion"
+    And I add pre-generated SSH public key to authorized_keys of host "sle_minion"
+  
+  Scenario: Bootstrap a SLES minion using SSH Key with wrong passphrase
```
"SSH key", lowercase.

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    And I enter "root" as "user"
+    And I check radio button "SSH Private Key"
+    And I attach the file "ssh_keypair/id_rsa_bootstrap-passphrase_linux" to "privKeyFile"
+    And I enter "you-shall-not-pass" as "privKeyPwd"
```
Do we have balrogs for our tests? ;-)

hustodemon replied:
Yes, and we need to make sure they don't pass ;)

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    And I click on "Bootstrap"
+    And I wait until I see "Permission denied, no authentication information" text
+
+  Scenario: Bootstrap a SLES minion using SSH Key
```
"key", lowercase

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    And I click on "Bootstrap"
+    And I wait until I see "Successfully bootstrapped host!" text
+
+  Scenario: Check new minion bootstrapped with SSH Key in System Overview page
```
same

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+  Scenario: Cleanup authorized keys
+    When I restore the SSH authorized_keys file of host "sle_minion"
+
+  Scenario: Cleanup: turn the SLES minion into a container build host after bootstrap with SSH Key
```
"key", lowercase. I will stop repeating it now, but there are more.

hustodemon replied:
I substituted all of them.

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+  Scenario: Cleanup: check that the minion is now a build host after bootstrap with SSH Key
+    Given I am on the Systems overview page of this "sle_minion"
+    Then I should see a "[Container Build Host]" text
+    Then I should see a "[OS Image Build Host]" text
```
"And"

Bischoff commented on testsuite/features/step_definitions/common_steps.rb:
```diff
@@ -1161,3 +1161,35 @@ def server_secret
+When(/^I backup the SSH authorized_keys file of host "([^"]*)"$/) do |host|
+  target = get_target(host)
+  _, ret_code = target.run("cp #{auth_keys_path} #{auth_keys_sav_path}")
+  raise 'error backing up authorized_keys on host' if ret_code.nonzero?
```
"Error", upper case

hustodemon replied:
Ok (although it seems the upper/lowercase convention is rather random in that file :) ).

Bischoff commented on testsuite/features/step_definitions/common_steps.rb:
```diff
@@ -1161,3 +1161,35 @@ def server_secret
+  target = get_target(host)
+  ret_code = file_inject(
+    target,
+    File.dirname(__FILE__) + '/../upload_files/ssh_keypair/' + key_filename,
```
or two string interpolations with `#{...}`(not sure it's a good idea, feel free to ignore)

hustodemon replied:
`common_steps` contain every possible way to combine path now, so I'll leave this for now since it works ;) (best way would probably be `File.join(File.dirname(__FILE__), 'something')`, but I don't want to break already working code).

Bischoff replied:
yes, it might be we have been creative here... :blush:

Bischoff commented on testsuite/features/secondary/sle_minion_bootstrap_ssh_key.feature:
```diff
@@ -0,0 +1,92 @@
+    Given I am on the Systems overview page of this "sle_minion"
+    Then I should see a "Default" text
+  
+  Scenario: Cleanup authorized keys
```
"Cleanup: restore authorized keys"

Bischoff commented:
> LGTM. Only few remarks, but nothing blocking the merge.
> 
> After merged, I would like to discuss with the rest of the squad to have it as a separate feature of this authentication.

I think this is already a separated feature. I think he just forgot to declare it in `run_sets`.

hustodemon commented:
Guys, thanks for you reviews!

Addressed all comments except of 2:
- the navigation step (`I navigate to "rhn/..."). I think this should be addressed globally - see my comment,
- the file joining code: again - see my inline comment.

I pushed the changes in separate commits, so that they can be easily reviewed.

Bischoff commented on testsuite/run_sets/secondary.yml:
```diff
@@ -17,6 +17,7 @@
 - features/secondary/trad_centos_client.feature
 - features/secondary/min_ubuntu_salt.feature
 - features/secondary/minssh_bootstrap_xmlrpc.feature
+- features/secondary/min_bootstrap_ssh_key.feature
```
thanks!
